ROX-18149:  byodb migration rocksdb transactions

dashrews78 opened the pull request:
## Description

Updates to wrap the RocksDB->Postgres migrations in a transaction.  This will enable use to recover from failures and to pick up with the last migration we successfully processed.  This is important when upgrading to a single database environment such as external or 4.2 where databases are not copied.  This provides us with a consistent state for the initial set of migrations.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
~- [ ] Evaluated and added CHANGELOG entry if required~
~- [ ] Determined and documented upgrade steps~
~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs]~(https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

Forced chaos scenarios by killing pods during migrations to ensure that data from the one being processed during the failure did not leak into the database AND that the subsequent restart picked up with the migration that was being processed when the failure occurred.

dashrews78 committed:
X-Smart-Branch-Parent: master

dashrews78 commented on migrator/clone/postgres/db_clone_manager_impl.go:
```diff
@@ -122,8 +122,6 @@ func (d *dbCloneManagerImpl) Scan() error {
 		log.Infof("clone %s is of version %v", PreviousClone, ver)
 
 		d.cloneMap[PreviousClone] = metadata.NewPostgres(ver, PreviousClone)
-	} else {
```
This was causing issue rolling back after a restart because it would delete previous.  Since dodging that would cause a fair amount of logic, I simply won't delete the `central_previous`.

c-du commented on migrator/runner/runner.go:
```diff
@@ -74,6 +77,18 @@ func Run(databases *types.Databases) error {
+		ctx = pgPkg.ContextWithTx(ctx, tx)
+
+		// Set the context with the databases so the wrapped transaction can be used
+		databases.DBCtx = ctx
```
It looks like we do not need to embed ctx into databases. How about passing it to migration.Run(...) as a parameter?

dashrews78 replied:
That would require a change to the interface which would ripple through all the `Run` functions.  I was trying to avoid that.

c-du commented on migrator/runner/runner.go:
```diff
@@ -84,11 +99,17 @@ func runMigrations(databases *types.Databases, startingSeqNum int) error {
+		}
+
+		// Outside transaction because of Gorm of course
+		err = updateVersion(databases, migration.VersionAfter)
```
Can we put this transaction together with the migration? Otherwise we should handle the case of re-entry here.

dashrews78 committed:
X-Smart-Squash: Squashed 14 commits:

477246abfb first pass
5cfc65148b took out the stupid artifact I left in
570d154063 style
46731f6c2f typo
ad8019928e lint
76a7d8b28c no truncate just transaction
527e506e00 update to pick up where leaving off
f1db768140 missed assignment
4e31724600 cleanup
18dba322a8 bug where a bounce would have deleted previous
e4f26652a0 minor cleanup before review
4977b87a64 style
b14635e6c7 update update version to use pgx instead of gorm
6314c110cd fix a nil context

dashrews78 committed:
figure out why pw making its way into the log

dashrews78 committed:
figure out why pw making its way into the log

dashrews78 committed:
trim space on password for compare

dashrews78 committed:
remove the logging
